### ISSUE TITLE
Causing Freezing/Device Lag

### ISSUE THREAD

**Issue by keeganjk** [ranger/ranger#2158]
#### Runtime Environment

`ranger version: ranger-master`
- Operating system and version: `5.9.6-artix1-1`
- Terminal emulator and version: `st 0.8.2`
- Python version: `Python version: 3.8.6 (default, Sep 30 2020, 20:59:51) [GCC 10.2.0]`
- Ranger version/commit: `Installed: 1.9.3.99.g240d8a0e-1`
- Locale: `Locale: en_US.UTF-8`

#### Current Behavior
When using `ranger`, especially when renaming files and navigating through the text in filenames, the device freezes up so that keyboard/mouse input does nothing and audio is put into a loop.

#### Expected Behavior
`ranger` behaving normally (simply navigating, renaming files) without slowing the device

#### Context

I have had to shut down my computer twice due what seem to be `ranger` freezing it up

#### Possible Solutions
Perhaps there is some sort of loop going on, or an inefficiency that makes it unable to handle holding down an arrow key to navigate through text (this is what I was doing the second time)

#### Steps to reproduce
1. Use `ranger`
2. Use `:rename` on a file
3. Move around in the text block

#### Traceback

I was unable to view any crashing of `ranger` itself

**Comment by toonn** [comment]
While this sounds unlikely there could be a bug no one's stumbled on so far. There's not enough info here to do much about it though. Please try to figure out a narrower cause. If we can't reproduce the issue it's nearly impossible to fix. I'm unable to reproduce this by moving the cursor back and forth in the console after using `A` on a file.

**Comment by keeganjk** [comment]
It seems to be a recent issue, there may be an underlying issue; I'm not sure.

**Comment by keeganjk** [comment]
It seems that the freezing issue is caused specifically by `scope.sh`, based on a `ps -fA` I run while experience lags.

[scope.sh](https://github.com/ranger/ranger/files/5908881/scope.txt)

**Comment by toonn** [comment]
I think scope is executed on the file before you even start renaming so I don't know why scope would only be freezing up ranger if you then try to rename the file.

**Comment by keeganjk** [comment]
It seemed to also be happening from viewing PDF/EPUB previews and also possibly from opening a shell in a directory within `ranger`.

**Comment by toonn** [comment]
I suspect this is just too many previews gumming up the works. Each file you move over spawns an instance of scope to generate a preview and you're simply spawning too many of them. Or you're spawning only a few but they're heavy.

**Comment by joeyshi12** [comment]
I'm also experiencing this problem when scrolling over large pdfs or within my downloads folder. My computer completely freezes whenever this happens

**Comment by toonn** [comment]
I've had thoughts about using a queue for preview generation so we don't spawn too many processes that end up never being looked at anyway. We could drop anything in the queue if the user leaves a directory for example. Not sure what else could help here other than just disabling problematic previews in scope.sh, of course.

**Comment by techdragon** [comment]
I'm getting this preview behaviour a lot when navigating directories full of text files. Not even big ones. Nothing over 128K, all plain text with `.txt`. Yet scrolling down will leave me with dozens and dozens of preview generation sub-processes that seem completely stalled (and don't respond to kill due to the blocking read issues mentioned elsewhere)

**Comment by joeyshi12** [comment]
> I've had thoughts about using a queue for preview generation so we don't spawn too many processes that end up never being looked at anyway. We could drop anything in the queue if the user leaves a directory for example. Not sure what else could help here other than just disabling problematic previews in scope.sh, of course.

Yes, disabling problematic previews (or really just gif previews) seemed to fix most of the issues I was dealing with. I have noticed freezing still occurs when I use ranger, but it's rare and I'm not sure how to consistently reproduce the bug anymore.
I added `image/gif) exit 1;;` in my `scope.sh#handle_image` for anyone also having issues with gifs: https://github.com/joeyshi12/dotfiles/blob/main/.config/ranger/scope.sh

**Comment by toonn** [comment]
This might be linked to #1787, a fix for that issue may potentially alleviate this one, if not fix it.

**Comment by YuCao16** [comment]
Hi @toonn, Thank you very much for your efforts, it was an AMAZING project.

I find that for Mac users this problem is still not solved. I have briefly summarised my experience of using ranger in the hope that it will help this project.

1. The first thing is that linux (eg. Ubuntu) really doesn't have this problem, if I hold down `j` to quickly browse a folder, the process will eventually end if I wait a while (weird for me). But for MacOS, the same folder, the process will not end.
2. My solution now on Mac is to set `use_preview_script` to `false`. When I need to preview, I then open it again by `zv` and use `zi` to preview the image, then in order to avoid freezing, I have to disable preview by pressing `zv` again.
3. Apart from render images, the main thing that causes freezing is big json files(#2524 and #2634), including `*.ipynb` files. I have a few `*.ipynb` files over 10mb that get stuck whenever I preview them.

I've thought of a few solutions, although I know this may be unnecessary (as it doesn't affect Linux users) / not practical  for you.
1. If it is possible to still use `zi` to preview images with `use_preview_script` set to `false`. This way I don't need to set this variable to true again.
2. It would be better to set a default file type/file size that does not toggle preview, e.g. {'*.json': '5mb', etc}, and then the preview could be triggered manually with a shortcut.
3. If a preview delay could be added, then when I press `j` to quickly browse a folder, there is no preview action acting on any file.
4. Adding a command to end the process when the cursor accidentally moves to a dangerous file (eg. 10mb json file), but this does not seem possible for Python.

Thanks again, Ranger is the best terminal files manager I have found so far!

**Comment by toonn** [comment]
@YuCao16, in your specific case it sounds like simply editing `scope.sh`, which is considered to be a configuration file, and removing the various previews you don't want or you can add `exit 2` to the cases so ranger just shows the plain text content.

**Comment by Ego2509** [comment]
Hello everyone, I believe this exact behavior is shown when changing names or previewing docx files that have hyperlinks. I guess thanks to this thread, disallowing previews on docx files helped me as I mostly don't need it.

OS: kali linux
Steps: renaming docx files in a folder with multiple docx files. It happens if your docx file has an undefined hyperlink.

Source: I made the broken docx file on purpose and the hyperlink requests to an external server.
